### ISSUE TITLE
Inconsistent results - Mittens VS standard Glove

### ISSUE THREAD

**Issue by olivierroncalez** [roamanalytics/mittens#10]
I am using mittens with a pre-built cooccurrence matrix of domains with the hopes of clustering certain domains that are thematically related, close to each other. Using the non-vectorized glove implementation from https://github.com/stanfordnlp/GloVe, I get very strong results. The current initialization is:

```
glove_model = Glove(no_components=50, learning_rate=0.03)
glove_model.fit(coo_matrix(matrix, dtype=float), epochs=50, no_threads=64, verbose=True)
```

Finding the nearest domains to `nintendo` using cosine distance yields good results.
```
find_nearest(glove_model, "nintendo", 10)
```

> [('game', 0.955347329117499),
>  ('zavvi', 0.9382098190168783),
>  ('eurogamer', 0.9296358002057901),
>  ('playstation', 0.9290108695965159),
>  ('gamespot', 0.9241452666014682),
>  ('gamesradar', 0.9210470827690169),
>  ('365games', 0.9193152241566838),
>  ('ign', 0.9178656620515147),
>  ('ea', 0.912055674280889),
>  ('forbiddenplanet', 0.9118661547211797)]

Given these results, I wanted to use `mittens` for two reasons: take advantage of the vectorized implementation for speed, and harness the ability to extend glove into a retrofitted model. However, when I used a basic mittens (without retrofitting existing embeddings), the results come out quite poor, even when the same hyperparameters are used.

```
glove_mittens_50_50 = GloVe(n=50, max_iter=50, learning_rate=0.03)
cooccurance = np.array(matrix.todense()) # was sparse matrix for original glove
glove_mittens_trained_50_50 = glove_mittens_50_50.fit(cooccurance)
```

I built a pd dataframe with the resulting numpy matrix and incorporated the domains as the index before writing a function that would calculate the cosine distance in the same way that the original glove model does. 

```
find_nearest(mittens_glove_df_50_50, "nintendo", 10)
```

> [('hmrc', 0.9992567),
>  ('anglingdirect', 0.999141),
>  ('axa', 0.99907136),
>  ('greatist', 0.99906415),
>  ('techadvisor', 0.99906313),
>  ('victorianplumbing', 0.99903136),
>  ('dell', 0.9990228),
>  ('imore', 0.99899846),
>  ('carpetright', 0.99899185)]

As you can see, the results are not at all as expected. Furthermore, while the original glove model will have converged and not change much (only very slightly) by increasing the number of iterations, the vectorized glove in this package will.

```
find_nearest(mittens_glove_df_50_100, "github", 10)  # 100 iterations
```

> [('yammer', 0.9993163),
>  ('twitch', 0.9992425),
>  ('axs', 0.9992203),
>  ('rottentomatoes', 0.99920493),
>  ('travelsupermarket', 0.99919695),
>  ('lbc', 0.99919665),
>  ('motors', 0.99918556),
>  ('goodreads', 0.9991843),
>  ('deezer', 0.9991767),
>  ('nationalexpress', 0.99917376)]

Is there a reason why this is the case? Am I doing anything wrong, or is there anything else you'd like me to try? 

Thanks.

**Comment by alexandru-dinu** [comment]
I am experiencing similar issues, but I am not certainly sure that my method is 100% correct so that to only suspect `mittens`. Here's what I've tried:

I have constructed a vocabulary consisting of the top 20,000 most frequent words from a custom corpus (python documentation [1]), and the corresponding co-occurrence matrix (20k x 20k), then I have fine-tuned `glove.840B.300d` embeddings as mentioned in the readme (with `window size = 5, num_iter = 5000`), thus obtaining a `20000 x 300` matrix.

The final embedding matrix was constructed as such:
```
final_glove_emb = load('glove.840B.300d')

for w in fine_tuned_glove_emb:
    if w not in glove_emb:
        final_glove_emb[w] = fine_tuned_glove_emb[w]
    else:
        final_glove_emb[w] = 0.5 * fine_tuned_glove_emb[w] + 0.5 * final_glove_emb[w]
```

Here's a list of 20 most similar words to `function`:
```
[('fromaddr', 0.6572332),
 ('functions', 0.65711826),
 ('preserving', 0.6557628),
 ('roundrobiniterables', 0.6226967),
 ('parameters', 0.59739476),
 ('asynciosleepdelay', 0.5884969),
 ('encodes', 0.5867052),
 ('preservation', 0.582687),
 ('maintaining', 0.58013296),
 ('singledispatch', 0.5575917),
 ('fractionsgcd', 0.5569468),
 ('preserve', 0.5410574),
 ('restoring', 0.53190446),
 ('snippet', 0.5317656),
 ('distutilscoresetup', 0.5273622),
 ('conserving', 0.51967645),
 ('essential', 0.50960046),
 ('defect', 0.5081374),
 ('subsection1', 0.5065669),
 ('vital', 0.50164175)]
```
I was hoping for words like `procedure`, `call`, `arguments` etc to appear at the top. 

I am guessing that maybe the python docs may not serve as a quality corpus for what I'm trying to achieve? I just want to rule out the possibility that there's something wrong with how I've performed the fine-tuning process, so that to focus only on the 'data' part.

Thanks.

[1] https://docs.python.org/3/download.html

**Comment by rajatcodes** [comment]
From the ReadMe file it says that we need pre-trained matrix in order to use mittens. This means you should have implied some kind of reweighting scheme to your co-occurrence matrix and then use Mittens.

**Comment by xjp08** [comment]
I met the same problem,and I think it's about how co-occurrence matrix generates.
